File: src/test/scala/com/codahale/cassie/tests/examples/CassieRun.scala
package com.codahale.cassie.tests.examples

import com.codahale.cassie._
import client.{RoundRobinHostSelector, PooledClientProvider, ClusterMap}
import clocks.MicrosecondEpochClock
import codecs.{VarInt, AsciiString, FixedLong}
import com.codahale.logula.Logging
import java.util.logging.Level

object CassieRun extends Logging {
  def main(args: Array[String]) {
    Logging.configure(Level.ALL)

    implicit val clock = MicrosecondEpochClock

    // pull down data about all the nodes in the cluster
    val map = new ClusterMap("localhost", 9160)

    // create a round-robin pool of 1-5 connections
    val selector = new RoundRobinHostSelector(map)
    val pool = new PooledClientProvider(selector, 1, 5, 10)

    // create a cluster
    val cluster = new Cluster(pool)

    // create a keyspace
    val keyspace = cluster.keyspace("Keyspace1")

    // create a column family
    val cass = keyspace.columnFamily[String, String]("Standard1")

    log.info("inserting some columns")
    cass.insert("yay for me", Column("name", "Coda"))
    cass.insert("yay for me", Column("motto", "Moar lean."))

    cass.insert("yay for you", Column("name", "Niki"))
    cass.insert("yay for you", Column("motto", "Told ya."))

    cass.insert("yay for us", Column("name", "Biscuit"))
    cass.insert("yay for us", Column("motto", "Mlalm."))

    cass.insert("yay for everyone", Column("name", "Louie"))
    cass.insert("yay for everyone", Column("motto", "Swish!"))

    log.info("getting a column: %s", cass.getColumn("yay for me", "name"))
    // Some(Column(name,Coda,1271789761374109))

    log.info("getting a column that doesn't exist: %s", cass.getColumn("yay for no one", "name"))
    // None

    log.info("getting a column that doesn't exist #2: %s", cass.getColumn("yay for no one", "oink"))
    // None

    log.info("getting a set of columns: %s", cass.getColumns("yay for me", Set("name", "motto")))
    // Map(motto -> Column(motto,Moar lean.,1271789761389735), name -> Column(name,Coda,1271789761374109))

    log.info("getting a whole row: %s", cass.getRow("yay for me"))
    // Map(motto -> Column(motto,Moar lean.,1271789761389735), name -> Column(name,Coda,1271789761374109))

    log.info("getting a column from a set of keys: %s", cass.multigetColumn(Set("yay for me", "yay for you"), "name"))
    // Map(yay for you -> Column(name,Niki,1271789761390785), yay for me -> Column(name,Coda,1271789761374109))

    log.info("getting a set of columns from a set of keys: %s", cass.multigetColumns(Set("yay for me", "yay for you"), Set("name", "motto")))
    // Map(yay for you -> Map(motto -> Column(motto,Told ya.,1271789761391366), name -> Column(name,Niki,1271789761390785)), yay for me -> Map(motto -> Column(motto,Moar lean.,1271789761389735), name -> Column(name,Coda,1271789761374109)))

    cass.getColumnAs[FixedLong, AsciiString]("key", 2)
    cass.insert("digits", Column[VarInt, VarInt](1, 300))

    log.info("Iterating!")
    for ((key, col) <- cass.rowIterator(2)) {
      log.info("Found: %s", col)
    }

    log.info("removing a column")
    cass.removeColumn("yay for me", "motto")

    log.info("removing a row")
    cass.removeRow("yay for me")

    log.info("Batching up some stuff")
    cass.batch() { cf =>
      cf.removeColumn("yay for you", "name")
      cf.removeColumns("yay for us", Set("name", "motto"))
      cf.insert("yay for nobody", Column("name", "Burt"))
      cf.insert("yay for nobody", Column("motto", "'S funny."))
      cf.insert("bits!", Column("motto", FixedLong(20019L)))
    }
  }
}

File: cassie-core/src/test/java/com/twitter/cassie/jtests/examples/CassieRun.java
package com.twitter.cassie.jtests.examples;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;

import com.twitter.cassie.*;
import com.twitter.cassie.clocks.MicrosecondEpochClock;
import com.twitter.cassie.types.*;
import com.twitter.cassie.codecs.*;
import com.twitter.util.Function2;
import com.twitter.util.Function;
import com.twitter.util.Future;

public final class CassieRun {
  public static <V> HashSet<V> Set(V... values) {
    return new HashSet<V>(Arrays.asList(values));
  }

  public static void info(Object o) {
    System.out.println(o);
  }

  public static void main(String[] args) throws Exception {
    // create a cluster with a single seed from which to map keyspaces
    Cluster cluster = new Cluster("localhost");

    // create a keyspace
    Keyspace keyspace = cluster.keyspace("Keyspace1")
      .retryAttempts(5)
      .requestTimeoutInMS(5000)
      .minConnectionsPerHost(1)
      .maxConnectionsPerHost(10)
      .removeAfterIdleForMS(60000)
      .connect();

    // create a column family
    ColumnFamily<String, String, String> cass = keyspace.columnFamily("Standard1", Utf8Codec.get(), Utf8Codec.get(), Utf8Codec.get());

    info("inserting some columns");
    //note that these calls are async, the apply() is where the waiting happens
    cass.insert("yay for me", cass.newColumn("name", "Coda")).apply();
    cass.insert("yay for me", cass.newColumn("motto", "Moar lean.")).apply();

    cass.insert("yay for you", cass.newColumn("name", "Niki")).apply();
    cass.insert("yay for you", cass.newColumn("motto", "Told ya.")).apply();

    cass.insert("yay for us", cass.newColumn("name", "Biscuit")).apply();
    cass.insert("yay for us", cass.newColumn("motto", "Mlalm.")).apply();

    cass.insert("yay for everyone", cass.newColumn("name", "Louie")).apply();
    cass.insert("yay for everyone", cass.newColumn("motto", "Swish!")).apply();

    info("getting a column: " + cass.getColumn("yay for me", "name").apply());
    info("getting a column that doesn't exist: " + cass.getColumn("yay for no one", "name").apply());
    info("getting a column that doesn't exist #2: " + cass.getColumn("yay for no one", "oink").apply());
    info("getting a set of columns: " + cass.getColumns("yay for me", Set("name", "motto")).apply());
    info("getting a whole row: " + cass.getRow("yay for me").apply());
    info("getting a column from a set of keys: " + cass.multigetColumn(Set("yay for me", "yay for you"), "name").apply());
    info("getting a set of columns from a set of keys: " + cass.multigetColumns(Set("yay for me", "yay for you"), Set("name", "motto")).apply());

    // drop some UUID sauce on things
    cass.keysAs(LexicalUUIDCodec.get()).insert(new LexicalUUID(cass.clock()), cass.newColumn("yay", "boo")).apply();
    cass.namesAs(LongCodec.get()).valuesAs(Utf8Codec.get()).getColumn("key", 2L).apply();
    cass.namesAs(IntCodec.get()).valuesAs(IntCodec.get())
        .insert("digits", cass.newColumn(1, 300)).apply();

    info("Iterating!");
    Future f = cass.rowsIteratee(2).foreach(new Function2<String, List<Column<String, String>>, Object>() {
      public Object apply(String key, List<Column<String,String>> columns) {
        info("Found: " + key);
        return null;
      }
    });

    f.apply();

    Future f2 = cass.columnsIteratee(2, "yay for me").foreach(new Function<Column<String, String>, Object>() {
      public Object apply(Column<String,String> column){
        info("Found Columns Iteratee: " + column);
        return null;
      }
    });

    f2.apply();

    info("removing a column");
    cass.removeColumn("yay for me", "motto").apply();

    info("removing a row");
    cass.removeRow("yay for me").apply();

    info("Batching up some stuff");
    cass.batch()
      .removeColumn("yay for you", "name")
      .removeColumns("yay for us", Set("name", "motto"))
      .insert("yay for nobody", cass.newColumn("name", "Burt"))
      .insert("yay for nobody", cass.newColumn("motto", "'S funny."))
      .execute().apply();

    info("Wrappin' up");
    keyspace.close();
  }
}
